examples/generative/Density estimation using RealNVP.py

```diff
@@ -0,0 +1,211 @@
+    s_layer_4 = keras.layers.Dense(
+        output_dim, activation="relu", kernel_regularizer=regularizers.l2(reg)
+    )(s_layer_3)
+    s_layer_5 = keras.layers.Dense(
+        2, activation="tanh", kernel_regularizer=regularizers.l2(reg)
+    )(s_layer_4)
+
+    return keras.Model(inputs=input, outputs=[s_layer_5, t_layer_5])
+
+
+Coupling(2).summary()
+"""
+## Real NVP
+"""
+
+
+class RealNVP(keras.Model):
+    def __init__(self, num_coupling_layers):
+        super(RealNVP, self).__init__()
+
+        self.num_coupling_layers = num_coupling_layers
+
+        # Distribution of the latent space.
+        self.distribution = tfp.distributions.MultivariateNormalDiag(
+            loc=[0.0, 0.0], scale_diag=[1.0, 1.0]
+        )
+
+        self.masks = np.array(
+            [[0, 1], [1, 0]] * (num_coupling_layers // 2), dtype="float32"
+        )
+
+        self.loss_tracker = keras.metrics.Mean(name="loss")
+
+        self.layers_list = [Coupling(2) for i in range(num_coupling_layers)]
+
+    def call(self, x, training=True):
+        log_det_inv = 0
+        direction = 1
+        if training:
+            direction = -1
+        for i in range(self.num_coupling_layers)[::direction]:
+            x_masked = x * self.masks[i]
+            reversed_mask = 1 - self.masks[i]
+            s, t = self.layers_list[i](x_masked)
+            s *= reversed_mask
+            t *= reversed_mask
+            gate = (direction - 1) / 2
+            x = (
+                reversed_mask
+                * (x * tf.exp(direction * s) + direction * t * tf.exp(gate * s))
+                + x_masked
+            )
+            log_det_inv += gate * tf.reduce_sum(s, [1])
+
+        return x, log_det_inv
+
+    # Log likelihood of the normal distribution plus the log determinant of the jacobian.
+
+    def log_loss(self, x):
+        y, logdet = self(x)
+        log_likelihood = self.distribution.log_prob(y) + logdet
+        return -tf.reduce_mean(log_likelihood)
+
+    def train_step(self, data):
+        with tf.GradientTape() as tape:
+
+            loss = self.log_loss(data)
+
+        g = tape.gradient(loss, self.trainable_variables)
+        self.optimizer.apply_gradients(zip(g, self.trainable_variables))
+        self.loss_tracker.update_state(loss)
+
+        return {"loss": self.loss_tracker.result()}
+
+    def test_step(self, data):
+        loss = self.log_loss(data)
+        self.loss_tracker.update_state(loss)
+
+        return {"loss": self.loss_tracker.result()}
+
+
+"""
+## Model training
+"""
+
+model = RealNVP(num_coupling_layers=6)
+
+model.compile(optimizer=keras.optimizers.Adam(learning_rate=0.0001))
+
+history = model.fit(
+    normalized_data, batch_size=256, epochs=300, verbose=2, validation_split=0.2
+)
+
+"""
+## Performance evaluation
+"""
+
+plt.figure(figsize=(15, 10))
+plt.plot(history.history["loss"])
+plt.title("model loss")
+plt.ylabel("loss")
+plt.xlabel("epoch")
+
+# From data to latent space.
+z, _ = model(normalized_data)
```

Comment:
Use `predict` here, since you don't need to differentiate through the call to `model()`

Reply:
Please update this